File: api/services/model_sources.py
"""Validation and readiness helpers for manifest-declared Hugging Face sources."""

from __future__ import annotations

import re
import unicodedata
from pathlib import Path
from typing import Any


_SAFE_ID = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]*$")
_WINDOWS_DEVICE = re.compile(
    r"^(?:con|prn|aux|nul|com[1-9]|lpt[1-9])(?:\..*)?$", re.IGNORECASE
)
_WINDOWS_UNSAFE = re.compile(r'[<>"|?*\x00-\x1f]')


def _portable_segment(value: str, field: str) -> str:
    if (
        not value
        or value in {".", ".."}
        or value.endswith((".", " "))
        or ":" in value
        or _WINDOWS_UNSAFE.search(value)
        or _WINDOWS_DEVICE.fullmatch(value)
    ):
        raise ValueError(f'{field} contains unsafe path segment "{value}"')
    return value


def safe_source_id(value: Any, field: str = "model source id") -> str:
    if (
        not isinstance(value, str)
        or not value
        or value != value.strip()
        or _SAFE_ID.fullmatch(value) is None
    ):
        raise ValueError(f"{field} must be a safe non-empty identifier")
    return _portable_segment(value, field)


def safe_relative_path(value: Any, field: str, *, allow_dot: bool = False) -> str:
    if not isinstance(value, str) or not value or value != value.strip():
        raise ValueError(f"{field} must be a non-empty relative path")
    if allow_dot and value == ".":
        return value
    if value == "." or value.startswith("/") or "\\" in value:
        raise ValueError(f"{field} must be a safe relative POSIX path")
    for part in value.split("/"):
        _portable_segment(part, field)
    return value


def _safe_prefix(value: Any, field: str) -> str:
    path = value[:-1] if isinstance(value, str) and value.endswith("/") else value
    safe_relative_path(path, field)
    return value


def _prefixes(value: Any, field: str) -> list[str] | None:
    if not isinstance(value, list):
        raise ValueError(f"{field} must be an array")
    return [_safe_prefix(entry, f"{field}[{index}]") for index, entry in enumerate(value)]


def _safe_repo_id(value: Any, field: str) -> str:
    if not isinstance(value, str) or not value or value != value.strip() or "\\" in value:
        raise ValueError(f"{field} must be a non-empty Hugging Face repository id")
    parts = value.split("/")
    if len(parts) > 2 or any(
        part in {"", ".", ".."} or _SAFE_ID.fullmatch(part) is None for part in parts
    ):
        raise ValueError(f"{field} is not a safe Hugging Face repository id")
    return value


def _safe_revision(value: Any, field: str) -> str | None:
    if value is None:
        return None
    if (
        not isinstance(value, str)
        or not value
        or value != value.strip()
        or value.startswith("/")
        or "\\" in value
        or "\0" in value
        or any(part in {"", ".", ".."} for part in value.split("/"))
    ):
        raise ValueError(f"{field} must be a safe non-empty revision")
    return value


def normalize_model_sources(
    node: dict[str, Any], *, field_name: str = "model_sources"
) -> list[dict[str, Any]] | None:
    """Validate only the new contract; legacy fields remain untouched."""
    if "model_sources" not in node:
        return None
    raw_sources = node["model_sources"]
    if not isinstance(raw_sources, list) or not raw_sources:
        raise ValueError(f"{field_name} must be a non-empty array")

    aliases: dict[str, str] = {}
    sources: list[dict[str, Any]] = []
    for index, raw in enumerate(raw_sources):
        field = f"{field_name}[{index}]"
        if not isinstance(raw, dict):
            raise ValueError(f"{field} must be an object")
        source_id = safe_source_id(raw.get("id"), f"{field}.id")
        alias = unicodedata.normalize("NFC", source_id).casefold()
        if alias in aliases:
            raise ValueError(
                f'model source ids "{aliases[alias]}" and "{source_id}" are not portable-unique'
            )
        aliases[alias] = source_id
        if raw.get("provider") != "huggingface":
            raise ValueError(f'{field}.provider must be "huggingface"')
        checks = raw.get("checks")
        if not isinstance(checks, list) or not checks:
            raise ValueError(f"{field}.checks must be a non-empty array")

        source: dict[str, Any] = {
            "id": source_id,
            "provider": "huggingface",
            "repo_id": _safe_repo_id(raw.get("repo_id"), f"{field}.repo_id"),
            "destination": safe_relative_path(
                raw.get("destination"), f"{field}.destination", allow_dot=True
            ),
            "checks": [
                safe_relative_path(check, f"{field}.checks[{check_index}]")
                for check_index, check in enumerate(checks)
            ],
        }
        revision = (
            _safe_revision(raw["revision"], f"{field}.revision")
            if "revision" in raw
            else None
        )
        if "revision" in raw and revision is None:
            raise ValueError(f"{field}.revision must be a safe non-empty revision")
        include = (
            _prefixes(raw["include_prefixes"], f"{field}.include_prefixes")
            if "include_prefixes" in raw
            else None
        )
        skip = (
            _prefixes(raw["skip_prefixes"], f"{field}.skip_prefixes")
            if "skip_prefixes" in raw
            else None
        )
        if revision is not None:
            source["revision"] = revision
        if include is not None:
            source["include_prefixes"] = include
        if skip is not None:
            source["skip_prefixes"] = skip
        sources.append(source)
    return sources


def validate_model_node_ids(nodes: list[dict[str, Any]]) -> None:
    """Managed node roots must be unique on case-insensitive filesystems too."""
    seen: set[str] = set()
    for node in nodes:
        if not isinstance(node, dict):
            raise ValueError("model node must be an object")
        if isinstance(node.get("id"), str) and node["id"].casefold() == "_shared":
            raise ValueError('model node id "_shared" is reserved')
        node_id = safe_source_id(node.get("id"), "model node id")
        alias = node_id.casefold()
        if alias in seen:
            raise ValueError(f'model node id "{node_id}" is not portable-unique')
        seen.add(alias)


def normalize_weight_groups(manifest: dict[str, Any]) -> list[dict[str, Any]] | None:
    if "weight_groups" not in manifest:
        return None
    raw_groups = manifest["weight_groups"]
    if not isinstance(raw_groups, list) or not raw_groups:
        raise ValueError("weight_groups must be a non-empty array")

    aliases: dict[str, str] = {}
    groups: list[dict[str, Any]] = []
    for index, raw in enumerate(raw_groups):
        field = f"weight_groups[{index}]"
        if not isinstance(raw, dict):
            raise ValueError(f"{field} must be an object")
        raw_group_id = raw.get("id")
        if isinstance(raw_group_id, str) and raw_group_id.casefold() == "_shared":
            raise ValueError(f'{field}.id uses the reserved identifier "_shared"')
        group_id = safe_source_id(raw_group_id, f"{field}.id")
        alias = unicodedata.normalize("NFC", group_id).casefold()
        if alias in aliases:
            raise ValueError(
                f'weight group ids "{aliases[alias]}" and "{group_id}" '
                "are not portable-unique"
            )
        aliases[alias] = group_id
        sources = normalize_model_sources(
            {"model_sources": raw.get("model_sources")},
            field_name=f"{field}.model_sources",
        )
        groups.append({"id": group_id, "model_sources": sources})
    return groups


def normalize_weight_group_references(
    node: dict[str, Any],
    groups: list[dict[str, Any]] | None,
    *,
    field_name: str = "weight_groups",
) -> list[str] | None:
    if "weight_groups" not in node:
        return None
    raw_refs = node["weight_groups"]
    if not isinstance(raw_refs, list) or not raw_refs:
        raise ValueError(f"{field_name} must be a non-empty array of weight group ids")

    available = {
        unicodedata.normalize("NFC", group["id"]).casefold(): group["id"]
        for group in groups or []
    }
    aliases: dict[str, str] = {}
    refs: list[str] = []
    for index, raw in enumerate(raw_refs):
        group_id = safe_source_id(raw, f"{field_name}[{index}]")
        alias = unicodedata.normalize("NFC", group_id).casefold()
        if alias in aliases:
            raise ValueError(
                f'weight group references "{aliases[alias]}" and "{group_id}" '
                "are not portable-unique"
            )
        aliases[alias] = group_id
        canonical = available.get(alias)
        if canonical is None:
            raise ValueError(
                f'{field_name}[{index}] references unknown weight group "{group_id}"'
            )
        refs.append(canonical)
    return refs


def _path_has_symlink(root: Path, candidate: Path) -> bool:
    root = root.absolute()
    candidate = candidate.absolute()
    try:
        relative = candidate.relative_to(root)
    except ValueError:
        return True
    current = root
    if current.exists() and current.is_symlink():
        return True
    for part in relative.parts:
        current /= part
        if current.exists() and current.is_symlink():
            return True
    return False


def resolve_model_root(models_dir: Path, model_id: str) -> Path:
    if not isinstance(model_id, str):
        raise ValueError("Model id must be a string")
    parts = model_id.split("/")
    if len(parts) != 2:
        raise ValueError("Model id must identify one extension node")
    extension_id = safe_source_id(parts[0], "extension id")
    if parts[1].casefold() == "_shared":
        raise ValueError('Model node id "_shared" is reserved')
    node_id = safe_source_id(parts[1], "model node id")
    root = models_dir.absolute()
    candidate = root / extension_id / node_id
    if _path_has_symlink(root, candidate):
        raise ValueError("Model path resolves through a symlink")
    try:
        candidate.resolve().relative_to(root.resolve())
    except ValueError as exc:
        raise ValueError("Model path escapes the models directory") from exc
    return candidate


def resolve_weight_group_root(models_dir: Path, extension_id: str, group_id: str) -> Path:
    safe_extension_id = safe_source_id(extension_id, "extension id")
    if isinstance(group_id, str) and group_id.casefold() == "_shared":
        raise ValueError('Weight group id "_shared" is reserved')
    safe_group_id = safe_source_id(group_id, "weight group id")
    root = models_dir.absolute()
    candidate = root / safe_extension_id / "_shared" / safe_group_id
    if _path_has_symlink(root, candidate):
        raise ValueError("Weight group path resolves through a symlink")
    try:
        candidate.resolve().relative_to(root.resolve())
    except ValueError as exc:
        raise ValueError("Weight group path escapes the models directory") from exc
    return candidate


def resolve_weight_storage_root(models_dir: Path, target_id: str) -> Path:
    if not isinstance(target_id, str):
        raise ValueError("Weight target id must be a string")
    parts = target_id.split("/")
    if len(parts) == 2:
        return resolve_model_root(models_dir, target_id)
    if len(parts) == 3 and parts[1] == "_shared":
        return resolve_weight_group_root(models_dir, parts[0], parts[2])
    raise ValueError(
        "Weight target id must identify one model node or extension weight group"
    )


def resolve_source_destination(models_dir: Path, model_id: str, destination: str) -> Path:
    model_root = resolve_model_root(models_dir, model_id)
    safe_destination = safe_relative_path(destination, "destination", allow_dot=True)
    candidate = model_root if safe_destination == "." else model_root.joinpath(*safe_destination.split("/"))
    if _path_has_symlink(model_root, candidate):
        raise ValueError("Source destination resolves through a symlink")
    return candidate


def resolve_source_destination_at_root(model_root: Path, destination: str) -> Path:
    safe_destination = safe_relative_path(destination, "destination", allow_dot=True)
    candidate = (
        model_root
        if safe_destination == "."
        else model_root.joinpath(*safe_destination.split("/"))
    )
    if _path_has_symlink(model_root, candidate):
        raise ValueError("Source destination resolves through a symlink")
    return candidate


def resolve_download_path(destination: Path, filename: str) -> Path:
    safe_filename = safe_relative_path(filename, "Hugging Face repository file")
    candidate = destination.joinpath(*safe_filename.split("/"))
    if _path_has_symlink(destination, candidate):
        raise ValueError("Download target resolves through a symlink")
    return candidate


def model_sources_are_downloaded(
    models_dir: Path, model_id: str, sources: list[dict[str, Any]]
) -> bool:
    try:
        model_root = resolve_model_root(models_dir, model_id)
        return model_sources_are_downloaded_at_root(model_root, sources)
    except (KeyError, OSError, TypeError, ValueError):
        return False


def model_sources_are_downloaded_at_root(
    model_root: Path, sources: list[dict[str, Any]]
) -> bool:
    try:
        if not model_root.is_dir():
            return False
        for source in sources:
            destination = resolve_source_destination_at_root(
                model_root, source["destination"]
            )
            if not destination.is_dir():
                return False
            for check in source["checks"]:
                candidate = resolve_download_path(destination, check)
                if (
                    not candidate.is_file()
                    or candidate.stat().st_size <= 0
                    or _path_has_symlink(model_root, candidate)
                ):
                    return False
        return bool(sources)
    except (KeyError, OSError, TypeError, ValueError):
        return False


def weight_group_sources_are_downloaded(
    models_dir: Path, extension_id: str, group: dict[str, Any]
) -> bool:
    try:
        return model_sources_are_downloaded_at_root(
            resolve_weight_group_root(models_dir, extension_id, group["id"]),
            group["model_sources"],
        )
    except (KeyError, OSError, TypeError, ValueError):
        return False


def validate_source_file_plan(
    sources: list[dict[str, Any]], files_by_source: dict[str, list[str]]
) -> None:
    """Reject cross-source aliases before the first file is written."""
    aliases: dict[str, tuple[str, str]] = {}
    for source in sources:
        source_id = source["id"]
        destination = source["destination"]
        source_files = {
            safe_relative_path(filename, f'model source "{source_id}" file')
            for filename in files_by_source[source_id]
        }
        missing_checks = [check for check in source["checks"] if check not in source_files]
        if missing_checks:
            raise ValueError(
                f'Model source "{source_id}" checks files excluded from its download plan: '
                + ", ".join(missing_checks)
            )
        for safe_filename in source_files:
            target = safe_filename if destination == "." else f"{destination}/{safe_filename}"
            for value in (target, f"{target}.part"):
                alias = unicodedata.normalize("NFC", value).casefold()
                for previous_alias, (previous_source, previous_target) in aliases.items():
                    if previous_source == source_id:
                        continue
                    if (
                        alias == previous_alias
                        or alias.startswith(f"{previous_alias}/")
                        or previous_alias.startswith(f"{alias}/")
                    ):
                        raise ValueError(
                            "Model sources have a portable target collision: "
                            f'"{previous_source}:{previous_target}" and "{source_id}:{value}"'
                        )
                aliases[alias] = (source_id, value)

File: electron/main/model-sources.ts
import { existsSync, lstatSync, readdirSync, statSync } from 'node:fs'
import { readdir, rm } from 'node:fs/promises'
import { isAbsolute, relative, resolve } from 'node:path'

export interface ModelSource {
  id: string
  provider: 'huggingface'
  repo_id: string
  revision?: string
  destination: string
  include_prefixes?: string[]
  skip_prefixes?: string[]
  checks: string[]
}

export interface ModelSourceNode {
  model_sources?: unknown
}

export interface ModelWeightGroup {
  id: string
  sources: ModelSource[]
}

export interface ModelWeightManifest {
  weight_groups?: unknown
}

export interface ModelWeightNode extends ModelSourceNode {
  weight_groups?: unknown
}

const SAFE_ID = /^[A-Za-z0-9][A-Za-z0-9._-]*$/
const WINDOWS_DEVICE = /^(?:con|prn|aux|nul|com[1-9]|lpt[1-9])(?:\..*)?$/i
const WINDOWS_UNSAFE = /[<>"|?*\u0000-\u001f]/

function portableSegment(value: string, field: string): string {
  if (
    !value
    || value === '.'
    || value === '..'
    || value.endsWith('.')
    || value.endsWith(' ')
    || value.includes(':')
    || WINDOWS_UNSAFE.test(value)
    || WINDOWS_DEVICE.test(value)
  ) {
    throw new Error(`${field} contains unsafe path segment "${value}"`)
  }
  return value
}

export function safeModelSourceId(value: unknown, field = 'model source id'): string {
  if (typeof value !== 'string' || !value || value !== value.trim() || !SAFE_ID.test(value)) {
    throw new Error(`${field} must be a safe non-empty identifier`)
  }
  return portableSegment(value, field)
}

export function safeModelRelativePath(value: unknown, field: string, allowDot = false): string {
  if (typeof value !== 'string' || !value || value !== value.trim()) {
    throw new Error(`${field} must be a non-empty relative path`)
  }
  if (allowDot && value === '.') return value
  if (value === '.' || value.startsWith('/') || value.includes('\\') || isAbsolute(value)) {
    throw new Error(`${field} must be a safe relative POSIX path`)
  }
  for (const part of value.split('/')) portableSegment(part, field)
  return value
}

function safePrefix(value: unknown, field: string): string {
  if (typeof value !== 'string') return safeModelRelativePath(value, field)
  const path = value.endsWith('/') ? value.slice(0, -1) : value
  safeModelRelativePath(path, field)
  return value
}

function optionalPrefixes(value: unknown, field: string): string[] | undefined {
  if (value === undefined) return undefined
  if (!Array.isArray(value)) throw new Error(`${field} must be an array`)
  return value.map((entry, index) => safePrefix(entry, `${field}[${index}]`))
}

function safeRepoId(value: unknown, field: string): string {
  if (typeof value !== 'string' || !value || value !== value.trim() || value.includes('\\')) {
    throw new Error(`${field} must be a non-empty Hugging Face repository id`)
  }
  const parts = value.split('/')
  if (parts.length > 2 || parts.some((part) => !SAFE_ID.test(part) || part === '.' || part === '..')) {
    throw new Error(`${field} is not a safe Hugging Face repository id`)
  }
  return value
}

function safeRevision(value: unknown, field: string): string | undefined {
  if (value === undefined) return undefined
  if (typeof value !== 'string' || !value || value !== value.trim() || value.startsWith('/') || value.includes('\\') || value.includes('\0')) {
    throw new Error(`${field} must be a safe non-empty revision`)
  }
  if (value.split('/').some((part) => !part || part === '.' || part === '..')) {
    throw new Error(`${field} must be a safe revision`)
  }
  return value
}

export function normalizeModelSources(
  node: ModelSourceNode,
  fieldName = 'model_sources',
): ModelSource[] | undefined {
  if (!Object.prototype.hasOwnProperty.call(node, 'model_sources')) return undefined
  if (!Array.isArray(node.model_sources) || node.model_sources.length === 0) {
    throw new Error(`${fieldName} must be a non-empty array`)
  }

  const seen = new Map<string, string>()
  return node.model_sources.map((raw, index) => {
    const field = `${fieldName}[${index}]`
    if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) {
      throw new Error(`${field} must be an object`)
    }
    const value = raw as Record<string, unknown>
    const id = safeModelSourceId(value.id, `${field}.id`)
    const alias = id.normalize('NFC').toLowerCase()
    const previous = seen.get(alias)
    if (previous) throw new Error(`model source ids "${previous}" and "${id}" are not portable-unique`)
    seen.set(alias, id)
    if (value.provider !== 'huggingface') throw new Error(`${field}.provider must be "huggingface"`)
    if (!Array.isArray(value.checks) || value.checks.length === 0) {
      throw new Error(`${field}.checks must be a non-empty array`)
    }

    const source: ModelSource = {
      id,
      provider: 'huggingface',
      repo_id: safeRepoId(value.repo_id, `${field}.repo_id`),
      destination: safeModelRelativePath(value.destination, `${field}.destination`, true),
      checks: value.checks.map((check, checkIndex) => (
        safeModelRelativePath(check, `${field}.checks[${checkIndex}]`)
      )),
    }
    const revision = safeRevision(value.revision, `${field}.revision`)
    const include = optionalPrefixes(value.include_prefixes, `${field}.include_prefixes`)
    const skip = optionalPrefixes(value.skip_prefixes, `${field}.skip_prefixes`)
    if (revision !== undefined) source.revision = revision
    if (include !== undefined) source.include_prefixes = include
    if (skip !== undefined) source.skip_prefixes = skip
    return source
  })
}

export function validateModelNodeIds(nodes: Array<{ id?: unknown }>): void {
  const seen = new Set<string>()
  for (const node of nodes) {
    if (typeof node?.id === 'string' && node.id.toLowerCase() === '_shared') {
      throw new Error('model node id "_shared" is reserved')
    }
    const id = safeModelSourceId(node?.id, 'model node id')
    const alias = id.toLowerCase()
    if (seen.has(alias)) throw new Error(`model node id "${id}" is not portable-unique`)
    seen.add(alias)
  }
}

export function normalizeWeightGroups(manifest: ModelWeightManifest): ModelWeightGroup[] | undefined {
  if (!Object.prototype.hasOwnProperty.call(manifest, 'weight_groups')) return undefined
  if (!Array.isArray(manifest.weight_groups) || manifest.weight_groups.length === 0) {
    throw new Error('weight_groups must be a non-empty array')
  }

  const seen = new Map<string, string>()
  return manifest.weight_groups.map((raw, index) => {
    const field = `weight_groups[${index}]`
    if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) {
      throw new Error(`${field} must be an object`)
    }
    const value = raw as Record<string, unknown>
    if (typeof value.id === 'string' && value.id.toLowerCase() === '_shared') {
      throw new Error(`${field}.id uses the reserved identifier "_shared"`)
    }
    const id = safeModelSourceId(value.id, `${field}.id`)
    const alias = id.normalize('NFC').toLowerCase()
    const previous = seen.get(alias)
    if (previous) throw new Error(`weight group ids "${previous}" and "${id}" are not portable-unique`)
    seen.set(alias, id)
    const sources = normalizeModelSources(
      { model_sources: value.model_sources },
      `${field}.model_sources`,
    )
    return { id, sources: sources! }
  })
}

export function normalizeWeightGroupReferences(
  node: ModelWeightNode,
  groups: ModelWeightGroup[] | undefined,
  fieldName = 'weight_groups',
): string[] | undefined {
  if (!Object.prototype.hasOwnProperty.call(node, 'weight_groups')) return undefined
  if (!Array.isArray(node.weight_groups) || node.weight_groups.length === 0) {
    throw new Error(`${fieldName} must be a non-empty array of weight group ids`)
  }
  const available = new Map((groups ?? []).map((group) => [group.id.normalize('NFC').toLowerCase(), group.id]))
  const seen = new Map<string, string>()
  return node.weight_groups.map((raw, index) => {
    const id = safeModelSourceId(raw, `${fieldName}[${index}]`)
    const alias = id.normalize('NFC').toLowerCase()
    const previous = seen.get(alias)
    if (previous) throw new Error(`weight group references "${previous}" and "${id}" are not portable-unique`)
    seen.set(alias, id)
    const canonical = available.get(alias)
    if (!canonical) throw new Error(`${fieldName}[${index}] references unknown weight group "${id}"`)
    return canonical
  })
}

function pathHasSymlink(root: string, candidate: string): boolean {
  const rootPath = resolve(root)
  const rel = relative(rootPath, resolve(candidate))
  if (rel === '..' || rel.startsWith('../') || rel.startsWith('..\\') || isAbsolute(rel)) return true
  let current = rootPath
  try {
    if (existsSync(current) && lstatSync(current).isSymbolicLink()) return true
    for (const part of rel.split(/[/\\]/).filter(Boolean)) {
      current = resolve(current, part)
      if (existsSync(current) && lstatSync(current).isSymbolicLink()) return true
    }
  } catch {
    return true
  }
  return false
}

export function resolveModelRoot(modelsDir: string, modelId: string): string {
  if (typeof modelId !== 'string') throw new Error('Model id must be a string')
  const parts = modelId.split('/')
  if (parts.length !== 2) throw new Error('Model id must identify one extension node')
  const extensionId = safeModelSourceId(parts[0], 'extension id')
  if (parts[1].toLowerCase() === '_shared') throw new Error('Model node id "_shared" is reserved')
  const nodeId = safeModelSourceId(parts[1], 'model node id')
  const root = resolve(modelsDir)
  const extensionRoot = resolveExtensionModelRoot(modelsDir, extensionId)
  const modelRoot = resolve(extensionRoot, nodeId)
  if (pathHasSymlink(root, modelRoot)) throw new Error('Model path resolves through a symlink')
  return modelRoot
}

export function resolveExtensionModelRoot(modelsDir: string, extensionId: string): string {
  const safeExtensionId = safeModelSourceId(extensionId, 'extension id')
  const root = resolve(modelsDir)
  const extensionRoot = resolve(root, safeExtensionId)
  if (pathHasSymlink(root, extensionRoot)) throw new Error('Extension model path resolves through a symlink')
  return extensionRoot
}

export function resolveWeightGroupRoot(modelsDir: string, extensionId: string, groupId: string): string {
  const safeExtensionId = safeModelSourceId(extensionId, 'extension id')
  if (typeof groupId === 'string' && groupId.toLowerCase() === '_shared') {
    throw new Error('Weight group id "_shared" is reserved')
  }
  const safeGroupId = safeModelSourceId(groupId, 'weight group id')
  const root = resolve(modelsDir)
  const extensionRoot = resolveExtensionModelRoot(modelsDir, safeExtensionId)
  const groupRoot = resolve(extensionRoot, '_shared', safeGroupId)
  if (pathHasSymlink(root, groupRoot)) throw new Error('Weight group path resolves through a symlink')
  return groupRoot
}

export function weightGroupTargetId(extensionId: string, groupId: string): string {
  const safeExtensionId = safeModelSourceId(extensionId, 'extension id')
  const safeGroupId = safeModelSourceId(groupId, 'weight group id')
  return `${safeExtensionId}/_shared/${safeGroupId}`
}

export function resolveWeightStorageRoot(modelsDir: string, targetId: string): string {
  if (typeof targetId !== 'string') throw new Error('Weight target id must be a string')
  const parts = targetId.split('/')
  if (parts.length === 2) return resolveModelRoot(modelsDir, targetId)
  if (parts.length === 3 && parts[1] === '_shared') {
    return resolveWeightGroupRoot(modelsDir, parts[0], parts[2])
  }
  throw new Error('Weight target id must identify one model node or extension weight group')
}

export function areModelSourcesDownloadedAtRoot(modelRoot: string, sources: ModelSource[]): boolean {
  try {
    if (!existsSync(modelRoot) || pathHasSymlink(modelRoot, modelRoot)) return false
    return sources.every((source) => {
      const destination = source.destination === '.'
        ? modelRoot
        : resolve(modelRoot, ...source.destination.split('/'))
      if (!existsSync(destination) || pathHasSymlink(modelRoot, destination)) return false
      return source.checks.every((check) => {
        const candidate = resolve(destination, ...check.split('/'))
        if (!existsSync(candidate) || pathHasSymlink(modelRoot, candidate)) return false
        try {
          const stat = statSync(candidate)
          return stat.isFile() && stat.size > 0
        } catch {
          return false
        }
      })
    })
  } catch {
    return false
  }
}

export function areModelSourcesDownloaded(modelsDir: string, modelId: string, sources: ModelSource[]): boolean {
  try {
    const modelRoot = resolveModelRoot(modelsDir, modelId)
    return areModelSourcesDownloadedAtRoot(modelRoot, sources)
  } catch {
    return false
  }
}

export function areWeightGroupSourcesDownloaded(
  modelsDir: string,
  extensionId: string,
  group: ModelWeightGroup,
): boolean {
  try {
    return areModelSourcesDownloadedAtRoot(
      resolveWeightGroupRoot(modelsDir, extensionId, group.id),
      group.sources,
    )
  } catch {
    return false
  }
}

export function modelHasLocalData(modelsDir: string, modelId: string): boolean {
  try {
    const modelRoot = resolveModelRoot(modelsDir, modelId)
    return existsSync(modelRoot) && readdirSync(modelRoot).length > 0
  } catch {
    return false
  }
}

export function weightStorageHasLocalData(modelsDir: string, targetId: string): boolean {
  try {
    const root = resolveWeightStorageRoot(modelsDir, targetId)
    return existsSync(root) && readdirSync(root).length > 0
  } catch {
    return false
  }
}

// Mirrors the backend's cancel cleanup (api/routers/model.py): only the in-progress
// `.part` files are removed, so completed sources already on disk survive a cancel.
export async function removePartialDownloadArtifacts(modelRoot: string): Promise<void> {
  if (!existsSync(modelRoot)) return
  const entries = await readdir(modelRoot, { recursive: true, withFileTypes: true })
  await Promise.all(
    entries
      .filter((entry) => entry.isFile() && entry.name.endsWith('.part'))
      .map((entry) => rm(resolve(entry.parentPath ?? modelRoot, entry.name), { force: true })),
  )
}
